File: packages/sdk/pact-ts/src/dpop.ts
import { randomBytes } from "node:crypto";

import { DpopSignError } from "./errors.ts";
import { signEd25519Message, sha256Hex } from "./invariants/crypto.ts";
import { canonicalizeJson } from "./invariants/json.ts";

/**
 * The schema identifier for DPoP proofs. Must match pact-kernel's DPOP_SCHEMA constant.
 */
export const DPOP_SCHEMA = "pact.dpop_proof.v1";

/**
 * The body of a DPoP proof. Field names use snake_case to match Rust/serde serialization.
 * Field order in canonical JSON is alphabetical (RFC 8785), which also matches serde's default.
 *
 * Fields (alphabetical order as they appear in canonical JSON):
 *   action_hash, agent_key, capability_id, issued_at, nonce, schema, tool_name, tool_server
 */
export interface DpopProofBody {
  action_hash: string;
  agent_key: string;
  capability_id: string;
  issued_at: number;
  nonce: string;
  schema: string;
  tool_name: string;
  tool_server: string;
}

/**
 * A signed DPoP proof. The signature is Ed25519 over the canonical JSON of the body.
 */
export interface DpopProof {
  body: DpopProofBody;
  signature: string;
}

/**
 * Parameters for signDpopProof.
 */
export interface SignDpopProofParams {
  capabilityId: string;
  toolServer: string;
  toolName: string;
  actionArgs: unknown;
  agentSeedHex: string;
  nonce?: string;
  issuedAt?: number;
}

/**
 * Sign a DPoP proof for a PACT tool invocation.
 *
 * The proof body is serialized as RFC 8785 canonical JSON before signing,
 * ensuring compatibility with pact-kernel's verify_dpop_proof.
 *
 * @throws DpopSignError if the agentSeedHex is invalid or signing fails.
 */
export function signDpopProof(params: SignDpopProofParams): DpopProof {
  const {
    capabilityId,
    toolServer,
    toolName,
    actionArgs,
    agentSeedHex,
    nonce = randomBytes(16).toString("hex"),
    issuedAt = Math.floor(Date.now() / 1000),
  } = params;

  try {
    // Compute action_hash: SHA-256 hex of canonical JSON of actionArgs
    const actionHash = sha256Hex(canonicalizeJson(actionArgs));

    // Derive agent public key from seed
    const { public_key_hex: agentKey, signature_hex: _unused } = signEd25519Message(
      "derive_key",
      agentSeedHex,
    );

    // Build body with fields in alphabetical order (matches canonical JSON and Rust serde)
    const body: DpopProofBody = {
      action_hash: actionHash,
      agent_key: agentKey,
      capability_id: capabilityId,
      issued_at: issuedAt,
      nonce,
      schema: DPOP_SCHEMA,
      tool_name: toolName,
      tool_server: toolServer,
    };

    // Sign canonical JSON of body
    const bodyCanonical = canonicalizeJson(body);
    const { signature_hex } = signEd25519Message(Buffer.from(bodyCanonical, "utf8"), agentSeedHex);

    return { body, signature: signature_hex };
  } catch (cause) {
    if (cause instanceof DpopSignError) {
      throw cause;
    }
    throw new DpopSignError("failed to sign DPoP proof", { cause });
  }
}

File: packages/sdk/chio-cpp/src/dpop.cpp
#include "chio/dpop.hpp"

#include <chrono>
#include <iomanip>
#include <random>
#include <sstream>

#include "chio/invariants.hpp"
#include "json.hpp"

namespace chio {
namespace {

std::string random_nonce() {
  std::random_device rng;
  std::ostringstream out;
  for (int i = 0; i < 16; ++i) {
    auto byte = static_cast<unsigned int>(rng() & 0xffU);
    out << std::hex << std::setw(2) << std::setfill('0') << byte;
  }
  return out.str();
}

std::uint64_t now_unix_secs() {
  const auto now = std::chrono::system_clock::now().time_since_epoch();
  return static_cast<std::uint64_t>(
      std::chrono::duration_cast<std::chrono::seconds>(now).count());
}

}  // namespace

std::string DpopProof::to_json() const {
  return "{\"body\":" + body_json + ",\"signature\":" + detail::quote(signature_hex) + "}";
}

Result<DpopProof> sign_dpop_proof(const DpopSignParams& params) {
  auto canonical_args = invariants::canonicalize_json(params.action_args_json);
  if (!canonical_args) {
    return Result<DpopProof>::failure(canonical_args.error());
  }
  auto action_hash = invariants::sha256_hex_utf8(canonical_args.value());
  if (!action_hash) {
    return Result<DpopProof>::failure(action_hash.error());
  }

  auto key_signature =
      invariants::sign_utf8_message_ed25519("derive_key", params.agent_seed_hex);
  if (!key_signature) {
    return Result<DpopProof>::failure(key_signature.error());
  }
  const auto agent_key = detail::extract_json_string_field(key_signature.value(), "public_key_hex");
  if (agent_key.empty()) {
    return Result<DpopProof>::failure(
        Error{ErrorCode::Protocol, "failed to derive DPoP agent public key"});
  }

  const auto nonce = params.nonce.empty() ? random_nonce() : params.nonce;
  const auto issued_at = params.issued_at == 0 ? now_unix_secs() : params.issued_at;

  std::ostringstream body;
  body << "{"
       << "\"action_hash\":" << detail::quote(action_hash.value()) << ","
       << "\"agent_key\":" << detail::quote(agent_key) << ","
       << "\"capability_id\":" << detail::quote(params.capability_id) << ","
       << "\"issued_at\":" << issued_at << ","
       << "\"nonce\":" << detail::quote(nonce) << ","
       << "\"schema\":\"chio.dpop_proof.v1\","
       << "\"tool_name\":" << detail::quote(params.tool_name) << ","
       << "\"tool_server\":" << detail::quote(params.tool_server) << "}";

  auto signed_body = invariants::sign_json_ed25519(body.str(), params.agent_seed_hex);
  if (!signed_body) {
    return Result<DpopProof>::failure(signed_body.error());
  }
  const auto signature = detail::extract_json_string_field(signed_body.value(), "signature_hex");
  if (signature.empty()) {
    return Result<DpopProof>::failure(
        Error{ErrorCode::Protocol, "failed to sign DPoP proof body"});
  }

  return Result<DpopProof>::success(DpopProof{body.str(), signature});
}

}  // namespace chio
